soroban-env-host/src/crypto/bn254.rs

```diff
@@ -0,0 +1,354 @@
+        self.bn254_affine_vec_from_vecobj::<BN254_G1_SERIALIZED_SIZE, G1Config>(
+            vp,
+            ContractCostType::Bls12381G1CheckPointOnCurve,
+            false, // G1 subgroup check is not necessary
+            "G1",
+        )
+    }
+
+    pub(crate) fn bn254_checked_g2_vec_from_vecobj(
+        &self,
+        vp: VecObject,
+    ) -> Result<Vec<G2Affine>, HostError> {
+        self.bn254_affine_vec_from_vecobj::<BN254_G2_SERIALIZED_SIZE, G2Config>(
+            vp,
+            ContractCostType::Bls12381G2CheckPointOnCurve,
+            true,
+            "G2",
+        )
+    }
+
+    pub(crate) fn bn254_fr_from_u256val(&self, sv: U256Val) -> Result<Fr, HostError> {
+        self.charge_budget(ContractCostType::Bn254FrFromU256, None)?;
+        let fr = if let Ok(small) = U256Small::try_from(sv) {
+            Fr::from_le_bytes_mod_order(&u64::from(small).to_le_bytes())
+        } else {
+            let obj: U256Object = sv.try_into()?;
+            self.visit_obj(obj, |u: &U256| {
+                Ok(Fr::from_le_bytes_mod_order(&u.to_le_bytes()))
+            })?
+        };
+        Ok(fr)
+    }
+
+    pub(crate) fn bn254_pairing_internal(
+        &self,
+        vp1: &Vec<G1Affine>,
+        vp2: &Vec<G2Affine>,
+    ) -> Result<PairingOutput<Bn254>, HostError> {
+        self.charge_budget(ContractCostType::Bn254Pairing, Some(vp1.len() as u64))?;
+        // check length requirements
+        if vp1.len() != vp2.len() || vp1.is_empty() {
+            return Err(self.bn254_err_invalid_input(
+                format!(
+                    "pairing: invalid input vector lengths ({}, {})",
+                    vp1.len(),
+                    vp2.len()
+                )
+                .as_str(),
+            ));
+        }
+        // This calls into `bn254::multi_miller_loop`, which delegates to
+        // `ark_ec::models::bn::BnConfig::multi_miller_loop` with the
+        // parameters defined in `ark_bn254`.
+        //
+        // Panic analysis:
+        //
+        // The following potential panic conditions could exist:
+        // 1. if two input vector lengths are not equal. There is a `zip_eq`
+        // which panics if the length of the two vectors are not equal. This is
+        // weeded out up front.
+        //
+        // 2. `coeffs.next().unwrap()`. This occurs when the algorithm Loops
+        // over pairs of `(a: G1Affine, b: G2Affine)`, converting them into
+        // `Vec<(G1Prepared, G2Preared::EllCoeff<Config>)>`, the latter contains
+        // three elements of Fp2. For each pair, the coeffs.next() can at most
+        // be called twice, when the bit being looped over in `Config::X` is
+        // set. So this panic cannot happen.
+        //
+        // 3. if any of the G1Affine point is infinity. The ell() function which
+        // calls p.xy().unwrap(), which is when the point is infinity. This
+        // condition also cannot happen because when the pairs are generated,
```

Comment:
I don't see this actually happening, how exactly do we filter these? I assume `bn254_affine_deserialize` could be returning an error  for the zero points, but that's not filtering. I can also see a test for pairing zero points and it seems like it works, but I don't understand how exactly. 
We also don't seem to have a test where there is a pair of zeros together with non-zero pairs (or at least I couldn't find it).

Reply:
Sorry this comment was supposed to be removed. I still need to do panic analysis.

Reply:
But I believe the zero filtering happens in the arkworks library.

Reply:
That's still a bit suspicious because we reject the empty input vectors, but they might be effectively treated as empty by the arkworks library during pairing and we don't do anything about that. I don't understand the use cases for this well enough, but the inconsistency is a bit concerning. Maybe it's fine though, I don't really understand the context this would be used in, but just wanted to make sure that we're at least intentional about this.

Reply:
That's fair, but I still don't fully understand the rationale for prohibiting empty input vectors - you still may end up pairing empty vectors if all they contain are infinity pairs. Could that be an issue?

Reply:
Rejecting empty inputs at the top level is just a specification choice (this also matches how evm specified it). Technically we could also let it be a no-op but if user passes in empty vectors it is likely an error so it makes sense to reject it. 
infinity points are still valid inputs for pairing (internal filtering is just an optimization). Pretty sure passing two infinity vectors should work without issue (it’s a good test case to add @sisuresh). We also meter them by  vector length treating them as regular points.